Empire/Empire.cpp

```diff
@@ -107,6 +112,36 @@ namespace {
         return retval;
     }
 
+    float CalculateNewStockpile(int empire_id, int stockpile_location_id, float starting_stockpile,
+        const std::map<std::set<int>, float>& available_pp,
+        const std::map<std::set<int>, float>& allocated_pp,
+        const std::map<std::set<int>, float>& allocated_stockpile_pp)
+    {
+        const Empire* empire = GetEmpire(empire_id);
+        if (!empire) {
+            ErrorLogger() << "CalculateStockpileContribution() passed null empire.  doing nothing.";
+            return 0.0f;
+        }
+        float stockpile_transfer_efficiency = std::min(1.0f,empire->GetMeter("METER_IMPERIAL_PP_TRANSFER_EFFICIENCY")->Current());      
+        float stockpile_used = boost::accumulate(allocated_stockpile_pp | boost::adaptors::map_values, 0.0f);
+        float new_contributions = 0.0f;
+        for (auto const& available_group: available_pp) {
+            auto alloc_it = allocated_pp.find(available_group.first);
+            float allocated_here = (alloc_it == allocated_pp.end()) ? 0.0f : alloc_it->second;
+            float excess_here = available_group.second - allocated_here;
+            if (excess_here < EPSILON)
+                continue;
+            // Transfer excess to stockpile
+            new_contributions += excess_here * stockpile_transfer_efficiency;
+            DebugLogger() << "allocated_here " << allocated_here;
```

Comment:
If retained in final form, please change to `TraceLogger`, combine uses where possible, and provide the iteration key.
In test run, I have 2000+ log lines for this function and uses on ln 330-332, within a single turn.